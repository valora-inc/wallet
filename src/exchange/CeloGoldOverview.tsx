import React from 'react'
import { Trans, WithTranslation } from 'react-i18next'
import { StyleSheet, Text, View } from 'react-native'
import CurrencyDisplay from 'src/components/CurrencyDisplay'
import { celoTokenBalanceSelector } from 'src/goldToken/selectors'
import useBalanceAutoRefresh from 'src/home/useBalanceAutoRefresh'
import { withTranslation } from 'src/i18n'
import useSelector from 'src/redux/useSelector'
import colors from 'src/styles/colors'
import fontStyles from 'src/styles/fonts'
import variables from 'src/styles/variables'
import { Currency } from 'src/utils/currencies'

interface OwnProps {
  testID: string
}

type Props = WithTranslation & OwnProps

export function CeloGoldOverview({ t, testID }: Props) {
  useBalanceAutoRefresh()
  const celoBalance = useSelector(celoTokenBalanceSelector)

  const celoBalanceAmount = celoBalance ? { value: celoBalance, currencyCode: Currency.Celo } : null

  return (
    <View style={styles.container} testID={testID}>
      <Text style={styles.title}>{t('yourGoldBalance')}</Text>
      <Text style={styles.balance} testID="CeloBalance">
        {celoBalanceAmount && <CurrencyDisplay amount={celoBalanceAmount} />}
      </Text>
      <Text style={styles.localBalance}>
        {celoBalanceAmount ? (
          <Trans i18nKey="equalToAmount">
            Equal to <CurrencyDisplay amount={celoBalanceAmount} showLocalAmount={true} />
          </Trans>
        ) : (
          t('loadingExchangeRate')
        )}
      </Text>
    </View>
  )
}

const styles = StyleSheet.create({
  container: {
    paddingHorizontal: variables.contentPadding,
    marginVertical: 16,
  },
  title: {
    ...fontStyles.h2,
    marginBottom: 8,
  },
  balance: {
    ...fontStyles.mediumNumber,
    color: colors.dark,
    marginBottom: 8,
  },
  localBalance: {
    ...fontStyles.regular,
    color: colors.gray4,
  },
})

export default withTranslation<Props>()(CeloGoldOverview)
