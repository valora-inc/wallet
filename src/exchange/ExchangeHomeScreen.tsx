import BigNumber from 'bignumber.js'
import React, { useMemo, useRef } from 'react'
import { useTranslation } from 'react-i18next'
import { StyleSheet, Text, View } from 'react-native'
import Animated from 'react-native-reanimated'
import { SafeAreaView } from 'react-native-safe-area-context'
import { CeloExchangeEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import Touchable from 'src/components/Touchable'
import CeloGoldHistoryChart from 'src/exchange/CeloGoldHistoryChart'
import CeloNewsFeed from 'src/exchange/CeloNewsFeed'
import { exchangeHistorySelector } from 'src/exchange/reducer'
import InfoIcon from 'src/icons/InfoIcon'
import { LocalCurrencyCode } from 'src/localCurrency/consts'
import { convertDollarsToLocalAmount } from 'src/localCurrency/convert'
import { getLocalCurrencyCode, usdToLocalCurrencyRateSelector } from 'src/localCurrency/selectors'
import DrawerTopBar from 'src/navigator/DrawerTopBar'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { default as useSelector } from 'src/redux/useSelector'
import DisconnectBanner from 'src/shared/DisconnectBanner'
import colors from 'src/styles/colors'
import fontStyles from 'src/styles/fonts'
import variables from 'src/styles/variables'
import { tokensBySymbolSelector } from 'src/tokens/selectors'
import { getLocalCurrencyDisplayValue } from 'src/utils/formatting'

function navigateToGuide() {
  ValoraAnalytics.track(CeloExchangeEvents.celo_home_info)
  navigate(Screens.GoldEducation)
}

function ExchangeHomeScreen() {
  function dollarsToLocal(amount: BigNumber.Value) {
    return convertDollarsToLocalAmount(amount, localCurrencyCode ? localExchangeRate : 1)
  }

  function displayLocalCurrency(amount: BigNumber.Value) {
    return getLocalCurrencyDisplayValue(amount, localCurrencyCode || LocalCurrencyCode.USD, true)
  }

  const scrollPosition = useRef(new Animated.Value(0)).current
  const onScroll = Animated.event([
    {
      nativeEvent: {
        contentOffset: {
          y: scrollPosition,
        },
      },
    },
  ])
  const headerOpacity = useMemo(
    () => ({
      opacity: scrollPosition.interpolate({
        inputRange: [0, 100],
        outputRange: [0, 1],
        extrapolate: Animated.Extrapolate.CLAMP,
      }),
    }),
    []
  )

  const { t } = useTranslation()

  const tokensBySymbol = useSelector(tokensBySymbolSelector)
  const localCurrencyCode = useSelector(getLocalCurrencyCode)
  const localExchangeRate = useSelector(usdToLocalCurrencyRateSelector)
  const exchangeHistory = useSelector(exchangeHistorySelector)

  const exchangeHistoryLength = exchangeHistory.aggregatedExchangeRates.length
  const lastKnownPriceUsd =
    tokensBySymbol.CGLD?.lastKnownPriceUsd ||
    (exchangeHistoryLength &&
      exchangeHistory.aggregatedExchangeRates[exchangeHistoryLength - 1].exchangeRate) ||
    new BigNumber(0)

  const currentGoldRateInLocalCurrency = dollarsToLocal(lastKnownPriceUsd)
  let rateChangeInPercentage, rateWentUp

  if (exchangeHistoryLength) {
    const oldestGoldRateInLocalCurrency = dollarsToLocal(
      exchangeHistory.aggregatedExchangeRates[0].exchangeRate
    )
    if (oldestGoldRateInLocalCurrency) {
      const rateChange = currentGoldRateInLocalCurrency?.minus(oldestGoldRateInLocalCurrency)
      rateChangeInPercentage = currentGoldRateInLocalCurrency
        ?.div(oldestGoldRateInLocalCurrency)
        .minus(1)
        .multipliedBy(100)

      rateWentUp = rateChange?.gt(0)
    }
  }

  return (
    <SafeAreaView testID="ExchangeHomeScreen" style={styles.background} edges={['top']}>
      <DrawerTopBar
        scrollPosition={scrollPosition}
        middleElement={
          <Animated.View testID="Header" style={[styles.header, headerOpacity]}>
            {currentGoldRateInLocalCurrency && (
              <Text testID="CeloPriceInLocalCurrency" style={styles.goldPriceCurrentValueHeader}>
                {getLocalCurrencyDisplayValue(
                  currentGoldRateInLocalCurrency,
                  LocalCurrencyCode.USD,
                  true
                )}
              </Text>
            )}
            {rateChangeInPercentage && (
              <Text
                testID="CeloPriceChange"
                style={rateWentUp ? styles.goldPriceWentUpHeader : styles.goldPriceWentDownHeader}
              >
                {rateWentUp ? '▴' : '▾'} {rateChangeInPercentage.toFormat(2)}%
              </Text>
            )}
          </Animated.View>
        }
      />
      <Animated.ScrollView
        scrollEventThrottle={16}
        onScroll={onScroll}
        // style={styles.background}
        testID="ExchangeScrollView"
        stickyHeaderIndices={[]}
        contentContainerStyle={styles.contentContainer}
      >
        <SafeAreaView style={styles.background} edges={['bottom']}>
          <DisconnectBanner />
          <View style={styles.goldPrice}>
            <View style={styles.goldPriceTitleArea}>
              <Text style={styles.goldPriceTitle}>{t('goldPrice')}</Text>
              <Touchable onPress={navigateToGuide} hitSlop={variables.iconHitslop}>
                <InfoIcon testID="ExchangeHomeScreen/Info" size={14} />
              </Touchable>
            </View>
            <View style={styles.goldPriceValues}>
              <Text style={styles.goldPriceCurrentValue}>
                {currentGoldRateInLocalCurrency
                  ? displayLocalCurrency(currentGoldRateInLocalCurrency)
                  : '-'}
              </Text>

              {rateChangeInPercentage && (
                <Text style={rateWentUp ? styles.goldPriceWentUp : styles.goldPriceWentDown}>
                  {rateWentUp ? '▴' : '▾'} {rateChangeInPercentage.toFormat(2)}%
                </Text>
              )}
            </View>
          </View>

          <CeloGoldHistoryChart testID="PriceChart" />
          <CeloNewsFeed />
        </SafeAreaView>
      </Animated.ScrollView>
    </SafeAreaView>
  )
}

export default ExchangeHomeScreen

const styles = StyleSheet.create({
  contentContainer: {
    flexGrow: 1,
    paddingBottom: variables.contentPadding,
  },
  header: {
    alignItems: 'center',
  },
  background: {
    flex: 1,
    justifyContent: 'space-between',
  },
  goldPrice: {
    padding: variables.contentPadding,
  },
  goldPriceTitleArea: {
    flexDirection: 'row',
    alignItems: 'center',
    marginBottom: 8,
    flexWrap: 'wrap',
  },
  goldPriceTitle: {
    ...fontStyles.h2,
    marginRight: 8,
  },
  goldPriceValues: {
    flexDirection: 'row',
    alignItems: 'flex-end',
    flexWrap: 'wrap',
  },
  goldPriceCurrentValue: {
    minHeight: 27,
    ...fontStyles.mediumNumber,
  },
  goldPriceCurrentValueHeader: {
    ...fontStyles.regular500,
  },
  goldPriceWentUp: {
    ...fontStyles.regular,
    color: colors.greenUI,
  },
  goldPriceWentDown: {
    ...fontStyles.regular,
    marginLeft: 4,
    color: colors.warning,
  },
  goldPriceWentUpHeader: {
    ...fontStyles.small600,
    color: colors.greenBrand,
  },
  goldPriceWentDownHeader: {
    ...fontStyles.small600,
    color: colors.warning,
  },
})
