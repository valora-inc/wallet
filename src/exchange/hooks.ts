import { exchangeRatesSelector } from 'src/exchange/reducer'
import useSelector from 'src/redux/useSelector'
import { Currency } from 'src/utils/currencies'
import { getRateForMakerToken } from 'src/utils/currencyExchange'

export function useDollarToCeloExchangeRate() {
  const exchangeRates = useSelector(exchangeRatesSelector)
  return getRateForMakerToken(exchangeRates, Currency.Celo, Currency.Dollar)
}
