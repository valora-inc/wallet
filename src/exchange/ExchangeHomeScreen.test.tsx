import { fireEvent, render } from '@testing-library/react-native'
import React from 'react'
import { Provider } from 'react-redux'
import ExchangeHomeScreen from 'src/exchange/ExchangeHomeScreen'
import { ExchangeRates } from 'src/exchange/reducer'
import { Screens } from 'src/navigator/Screens'
import { Currency } from 'src/utils/currencies'
import { createMockStore, getMockStackScreenProps } from 'test/utils'
import { makeExchangeRates } from 'test/values'

// Mock this for now, as we get apollo issues
jest.mock('src/transactions/TransactionsList', () => 'TransactionsList')

const mockScreenProps = getMockStackScreenProps(Screens.ExchangeHomeScreen)
const exchangeRates: ExchangeRates = makeExchangeRates('0.11', '10')

describe('ExchangeHomeScreen', () => {
  it('renders and behaves correctly for non CP-DOTO restricted countries', () => {
    const store = createMockStore({
      goldToken: { balance: '2' },
      stableToken: { balances: { [Currency.Dollar]: '10' } },
      exchange: { exchangeRates },
    })

    const tree = render(
      <Provider store={store}>
        <ExchangeHomeScreen {...mockScreenProps} />
      </Provider>
    )

    expect(tree).toMatchSnapshot()

    jest.clearAllMocks()
    fireEvent.press(tree.getByTestId('BuyCelo'))
    expect(mockScreenProps.navigation.navigate).toHaveBeenCalledWith(Screens.ExchangeTradeScreen, {
      buyCelo: true,
    })

    jest.clearAllMocks()
    fireEvent.press(tree.getByTestId('SellCelo'))
    expect(mockScreenProps.navigation.navigate).toHaveBeenCalledWith(Screens.ExchangeTradeScreen, {
      buyCelo: false,
    })

    jest.clearAllMocks()
    fireEvent.press(tree.getByTestId('WithdrawCELO'))
    expect(mockScreenProps.navigation.navigate).toHaveBeenCalledWith(Screens.WithdrawCeloScreen, {
      isCashOut: false,
    })
  })

  it('renders and behaves correctly for CP-DOTO restricted countries', () => {
    const store = createMockStore({
      networkInfo: {
        userLocationData: {
          countryCodeAlpha2: 'PH', // PH is restricted for CP-DOTO
          region: null,
          ipAddress: null,
        },
      },
      goldToken: { balance: '2' },
      stableToken: { balances: { [Currency.Dollar]: '10' } },
      exchange: { exchangeRates },
    })

    const tree = render(
      <Provider store={store}>
        <ExchangeHomeScreen {...mockScreenProps} />
      </Provider>
    )

    expect(tree).toMatchSnapshot()

    // Check we cannot buy/sell
    expect(tree.queryByTestId('BuyCelo')).toBeFalsy()
    expect(tree.queryByTestId('SellCelo')).toBeFalsy()

    // Check we can withdraw
    fireEvent.press(tree.getByTestId('WithdrawCELO'))
    expect(mockScreenProps.navigation.navigate).toHaveBeenCalledWith(Screens.WithdrawCeloScreen, {
      isCashOut: false,
    })
  })
})
