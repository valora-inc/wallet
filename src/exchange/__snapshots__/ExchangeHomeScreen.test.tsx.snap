// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExchangeHomeScreen renders and behaves correctly for CP-DOTO restricted countries 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "top",
    ]
  }
  style={
    Object {
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "borderBottomColor": 0,
        "borderBottomWidth": 1,
        "flexDirection": "row",
        "flexShrink": 1,
        "justifyContent": "center",
        "minHeight": 56,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      hitSlop={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "left": 0,
            "marginBottom": 0,
            "marginLeft": 16,
            "padding": 0,
            "position": "absolute",
          },
          Object {
            "opacity": 1,
          },
        ]
      }
    >
      <svg
        fill="none"
        height="32"
        testID="Hamburger"
        viewBox="0 0 32 32"
        width="32"
      >
        <line
          stroke="#2E3338"
          strokeLinecap="round"
          strokeWidth="2.5"
          x1="7.25"
          x2="24.75"
          y1="9.75"
          y2="9.75"
        />
        <line
          stroke="#2E3338"
          strokeLinecap="round"
          strokeWidth="2.5"
          x1="7.25"
          x2="24.75"
          y1="15.75"
          y2="15.75"
        />
        <line
          stroke="#2E3338"
          strokeLinecap="round"
          strokeWidth="2.5"
          x1="7.25"
          x2="24.75"
          y1="21.75"
          y2="21.75"
        />
      </svg>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
          },
          Object {
            "opacity": AnimatedValue {
              " __value": 0,
            },
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#2E3338",
            "fontFamily": "Inter-Medium",
            "fontSize": 16,
            "lineHeight": 22,
          }
        }
      >
        $9.09
      </Text>
    </View>
  </View>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "flexGrow": 1,
        "paddingBottom": 16,
      }
    }
    scrollEventThrottle={16}
    stickyHeaderIndices={Array []}
    testID="ExchangeScrollView"
  >
    <RNCSafeAreaView
      edges={
        Array [
          "bottom",
        ]
      }
      style={
        Object {
          "flex": 1,
          "justifyContent": "space-between",
        }
      }
    >
      <View
        style={
          Object {
            "padding": 16,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "flexWrap": "wrap",
              "marginBottom": 8,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#2E3338",
                "fontFamily": "Jost-Medium",
                "fontSize": 22,
                "lineHeight": 28,
                "marginRight": 8,
              }
            }
          >
            goldPrice
          </Text>
          <View
            accessible={true}
            focusable={true}
            hitSlop={
              Object {
                "bottom": 10,
                "left": 10,
                "right": 10,
                "top": 10,
              }
            }
            nativeBackgroundAndroid={
              Object {
                "attribute": "selectableItemBackground",
                "rippleRadius": undefined,
                "type": "ThemeAttrAndroid",
              }
            }
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <svg
              fill="none"
              height={14}
              viewBox="0 0 16 16"
              width={14}
            >
              <circle
                cx="8"
                cy="8"
                r="7"
                stroke="#2E3338"
                strokeWidth="1.25"
              />
              <path
                d="M8 12V7M8 6V5"
                stroke="#2E3338"
                strokeWidth="1.25"
              />
            </svg>
          </View>
        </View>
        <View
          style={
            Object {
              "alignItems": "flex-end",
              "flexDirection": "row",
              "flexWrap": "wrap",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#2E3338",
                "fontFamily": "Inter-Regular",
                "fontSize": 24,
                "lineHeight": 27,
                "minHeight": 27,
              }
            }
          >
            $9.09
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 217.5,
            "justifyContent": "center",
            "width": 750,
          }
        }
      >
        <ActivityIndicator
          color="#EEB93C"
          size="large"
        />
      </View>
      <View
        style={
          Array [
            Object {
              "flexDirection": "column",
            },
            Object {
              "marginBottom": 16,
              "marginHorizontal": 16,
              "marginTop": 24,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 100,
              },
            ]
          }
        >
          <View
            accessible={true}
            focusable={true}
            nativeBackgroundAndroid={
              Object {
                "attribute": "selectableItemBackground",
                "rippleRadius": undefined,
                "type": "ThemeAttrAndroid",
              }
            }
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#EEB93C",
                "flexGrow": 1,
                "height": 48,
                "justifyContent": "center",
                "opacity": undefined,
                "paddingHorizontal": 24,
                "paddingVertical": 5,
              }
            }
            testID="WithdrawCELO"
          >
            <Text
              maxFontSizeMultiplier={1}
              style={
                Object {
                  "color": "#FFFFFF",
                  "fontFamily": "Inter-SemiBold",
                  "fontSize": 16,
                  "lineHeight": 22,
                }
              }
            >
              withdrawCelo
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#EDEEEF",
            "height": 1,
            "marginHorizontal": 16,
          }
        }
      />
      <View
        style={
          Object {
            "marginVertical": 16,
            "paddingHorizontal": 16,
          }
        }
        testID="ExchangeAccountOverview"
      >
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Jost-Medium",
              "fontSize": 22,
              "lineHeight": 28,
              "marginBottom": 8,
            }
          }
        >
          yourGoldBalance
        </Text>
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Inter-Regular",
              "fontSize": 24,
              "lineHeight": 27,
              "marginBottom": 8,
            }
          }
          testID="CeloBalance"
        >
          <Text
            style={
              Array [
                undefined,
                Object {
                  "color": undefined,
                },
              ]
            }
            testID="undefined/value"
          >
            
            
            2.000
          </Text>
        </Text>
        <Text
          style={
            Object {
              "color": "#9CA4A9",
              "fontFamily": "Inter-Regular",
              "fontSize": 16,
              "lineHeight": 22,
            }
          }
        >
          Equal to 
          <Text
            style={
              Array [
                undefined,
                Object {
                  "color": undefined,
                },
              ]
            }
            testID="undefined/value"
          >
            
            â‚±
            6.00
          </Text>
        </Text>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#EDEEEF",
            "height": 1,
            "marginHorizontal": 16,
          }
        }
      />
      <View
        style={
          Object {
            "paddingBottom": 8,
            "paddingHorizontal": 16,
            "paddingTop": 20,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Jost-Medium",
              "fontSize": 22,
              "lineHeight": 28,
            }
          }
        >
          activity
        </Text>
      </View>
      <TransactionsList
        feedType="exchange"
      />
    </RNCSafeAreaView>
  </RCTScrollView>
</RNCSafeAreaView>
`;

exports[`ExchangeHomeScreen renders and behaves correctly for non CP-DOTO restricted countries 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "top",
    ]
  }
  style={
    Object {
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "borderBottomColor": 0,
        "borderBottomWidth": 1,
        "flexDirection": "row",
        "flexShrink": 1,
        "justifyContent": "center",
        "minHeight": 56,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      hitSlop={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "left": 0,
            "marginBottom": 0,
            "marginLeft": 16,
            "padding": 0,
            "position": "absolute",
          },
          Object {
            "opacity": 1,
          },
        ]
      }
    >
      <svg
        fill="none"
        height="32"
        testID="Hamburger"
        viewBox="0 0 32 32"
        width="32"
      >
        <line
          stroke="#2E3338"
          strokeLinecap="round"
          strokeWidth="2.5"
          x1="7.25"
          x2="24.75"
          y1="9.75"
          y2="9.75"
        />
        <line
          stroke="#2E3338"
          strokeLinecap="round"
          strokeWidth="2.5"
          x1="7.25"
          x2="24.75"
          y1="15.75"
          y2="15.75"
        />
        <line
          stroke="#2E3338"
          strokeLinecap="round"
          strokeWidth="2.5"
          x1="7.25"
          x2="24.75"
          y1="21.75"
          y2="21.75"
        />
      </svg>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
          },
          Object {
            "opacity": AnimatedValue {
              " __value": 0,
            },
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#2E3338",
            "fontFamily": "Inter-Medium",
            "fontSize": 16,
            "lineHeight": 22,
          }
        }
      >
        $9.09
      </Text>
    </View>
  </View>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "flexGrow": 1,
        "paddingBottom": 16,
      }
    }
    scrollEventThrottle={16}
    stickyHeaderIndices={Array []}
    testID="ExchangeScrollView"
  >
    <RNCSafeAreaView
      edges={
        Array [
          "bottom",
        ]
      }
      style={
        Object {
          "flex": 1,
          "justifyContent": "space-between",
        }
      }
    >
      <View
        style={
          Object {
            "padding": 16,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "flexWrap": "wrap",
              "marginBottom": 8,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#2E3338",
                "fontFamily": "Jost-Medium",
                "fontSize": 22,
                "lineHeight": 28,
                "marginRight": 8,
              }
            }
          >
            goldPrice
          </Text>
          <View
            accessible={true}
            focusable={true}
            hitSlop={
              Object {
                "bottom": 10,
                "left": 10,
                "right": 10,
                "top": 10,
              }
            }
            nativeBackgroundAndroid={
              Object {
                "attribute": "selectableItemBackground",
                "rippleRadius": undefined,
                "type": "ThemeAttrAndroid",
              }
            }
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <svg
              fill="none"
              height={14}
              viewBox="0 0 16 16"
              width={14}
            >
              <circle
                cx="8"
                cy="8"
                r="7"
                stroke="#2E3338"
                strokeWidth="1.25"
              />
              <path
                d="M8 12V7M8 6V5"
                stroke="#2E3338"
                strokeWidth="1.25"
              />
            </svg>
          </View>
        </View>
        <View
          style={
            Object {
              "alignItems": "flex-end",
              "flexDirection": "row",
              "flexWrap": "wrap",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#2E3338",
                "fontFamily": "Inter-Regular",
                "fontSize": 24,
                "lineHeight": 27,
                "minHeight": 27,
              }
            }
          >
            $9.09
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 217.5,
            "justifyContent": "center",
            "width": 750,
          }
        }
      >
        <ActivityIndicator
          color="#EEB93C"
          size="large"
        />
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginBottom": 28,
            "marginHorizontal": 12,
            "marginTop": 24,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "column",
              },
              Object {
                "flex": 1,
                "marginHorizontal": 4,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 100,
                },
              ]
            }
          >
            <View
              accessible={true}
              focusable={true}
              nativeBackgroundAndroid={
                Object {
                  "attribute": "selectableItemBackground",
                  "rippleRadius": undefined,
                  "type": "ThemeAttrAndroid",
                }
              }
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#EEB93C",
                  "flexGrow": 1,
                  "height": 48,
                  "justifyContent": "center",
                  "opacity": undefined,
                  "paddingHorizontal": 24,
                  "paddingVertical": 5,
                }
              }
              testID="BuyCelo"
            >
              <Text
                maxFontSizeMultiplier={1}
                style={
                  Object {
                    "color": "#FFFFFF",
                    "fontFamily": "Inter-SemiBold",
                    "fontSize": 16,
                    "lineHeight": 22,
                  }
                }
              >
                buy
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "flexDirection": "column",
              },
              Object {
                "flex": 1,
                "marginHorizontal": 4,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 100,
                },
              ]
            }
          >
            <View
              accessible={true}
              focusable={true}
              nativeBackgroundAndroid={
                Object {
                  "attribute": "selectableItemBackground",
                  "rippleRadius": undefined,
                  "type": "ThemeAttrAndroid",
                }
              }
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#EEB93C",
                  "flexGrow": 1,
                  "height": 48,
                  "justifyContent": "center",
                  "opacity": undefined,
                  "paddingHorizontal": 24,
                  "paddingVertical": 5,
                }
              }
              testID="SellCelo"
            >
              <Text
                maxFontSizeMultiplier={1}
                style={
                  Object {
                    "color": "#FFFFFF",
                    "fontFamily": "Inter-SemiBold",
                    "fontSize": 16,
                    "lineHeight": 22,
                  }
                }
              >
                sell
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#EDEEEF",
            "height": 1,
            "marginHorizontal": 16,
          }
        }
      />
      <View
        style={
          Object {
            "marginVertical": 16,
            "paddingHorizontal": 16,
          }
        }
        testID="ExchangeAccountOverview"
      >
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Jost-Medium",
              "fontSize": 22,
              "lineHeight": 28,
              "marginBottom": 8,
            }
          }
        >
          yourGoldBalance
        </Text>
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Inter-Regular",
              "fontSize": 24,
              "lineHeight": 27,
              "marginBottom": 8,
            }
          }
          testID="CeloBalance"
        >
          <Text
            style={
              Array [
                undefined,
                Object {
                  "color": undefined,
                },
              ]
            }
            testID="undefined/value"
          >
            
            
            2.000
          </Text>
        </Text>
        <Text
          style={
            Object {
              "color": "#9CA4A9",
              "fontFamily": "Inter-Regular",
              "fontSize": 16,
              "lineHeight": 22,
            }
          }
        >
          Equal to 
          <Text
            style={
              Array [
                undefined,
                Object {
                  "color": undefined,
                },
              ]
            }
            testID="undefined/value"
          >
            
            â‚±
            6.00
          </Text>
        </Text>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#EDEEEF",
            "height": 1,
            "marginHorizontal": 16,
          }
        }
      />
      <View
        style={
          Object {
            "backgroundColor": "#FFFFFF",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          nativeBackgroundAndroid={
            Object {
              "attribute": "selectableItemBackgroundBorderless",
              "rippleRadius": undefined,
              "type": "ThemeAttrAndroid",
            }
          }
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          testID="WithdrawCELO"
        >
          <View
            style={
              Object {
                "borderBottomColor": "#EDEEEF",
                "borderBottomWidth": 1,
                "marginLeft": 16,
                "paddingVertical": 16,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "paddingRight": 16,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "flex": 1,
                      "justifyContent": "center",
                    },
                  ]
                }
              >
                <Text
                  style={
                    Object {
                      "color": "#2E3338",
                      "fontFamily": "Inter-Regular",
                      "fontSize": 16,
                      "lineHeight": 22,
                    }
                  }
                >
                  withdrawCelo
                </Text>
              </View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                  }
                }
              >
                <svg
                  fill="none"
                  height={16}
                  style={Object {}}
                  testID="ForwardChevron"
                  viewBox="0 0 8 16"
                  width={8}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clipRule="evenodd"
                    d="M.293.293a1 1 0 0 1 1.414 0l6 6a1 1 0 0 1 0 1.414l-6 6a1 1 0 0 1-1.414-1.414L5.586 7 .293 1.707a1 1 0 0 1 0-1.414Z"
                    fill="#2E3338"
                    fillRule="evenodd"
                    style={Object {}}
                  />
                </svg>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "paddingBottom": 8,
            "paddingHorizontal": 16,
            "paddingTop": 20,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Jost-Medium",
              "fontSize": 22,
              "lineHeight": 28,
            }
          }
        >
          activity
        </Text>
      </View>
      <TransactionsList
        feedType="exchange"
      />
    </RNCSafeAreaView>
  </RCTScrollView>
</RNCSafeAreaView>
`;
